2-0-data-structures-and-algorithms/src/test/java/com/bobocode/bst/BinarySearchTreeTest.java

```diff
@@ -1,82 +1,249 @@
 
     @Test
     @Order(1)
-    void createWithElements() {
-        BinarySearchTree<Integer> bst = RecursiveBinarySearchTree.of(someElements);
-        for (var e: someElements) {
-            assertThat(bst.contains(e)).isTrue();
-        }
-        assertThat(bst.size()).isEqualTo(someElements.length);
+    void checkProperNodeClassName() {
+        tree = new RecursiveBinarySearchTree<>();
+        Class<?> innerClass = getInnerClass();
+        String name = innerClass.getSimpleName();
+
+        assertThat(name).isEqualTo("Node");
     }
 
     @Test
     @Order(2)
-    void insertUniqueElements() {
-        BinarySearchTree<Integer> bst = RecursiveBinarySearchTree.of();
-        for (int i = 0; i < someElements.length; i++) {
-            var e = someElements[i];
-            assertThat(bst.contains(e)).isFalse(); //does not contain
-            assertThat(bst.size()).isEqualTo(i);
+    void checkProperTreeFields() {
+        tree = new RecursiveBinarySearchTree<>();
+        Class<?> treeClass = tree.getClass();
 
-            assertThat(bst.insert(e)).isTrue(); //do insert
+        boolean hasSizeField = Arrays.stream(treeClass.getDeclaredFields())
+                .anyMatch(SIZE_FIELD);
 
-            assertThat(bst.contains(e)).isTrue(); //and contains
-            assertThat(bst.size()).isEqualTo(i + 1);
-        }
+        boolean hasNodeField = Arrays.stream(treeClass.getDeclaredFields())
+                .anyMatch(NODE_FIELD);
+
+        assertThat(hasSizeField).isTrue();
+        assertThat(hasNodeField).isTrue();
     }
 
     @Test
     @Order(3)
-    void insertNonUniqueElements() {
-        BinarySearchTree<Integer> bst = RecursiveBinarySearchTree.of(someElements);
-        for (var e: someElements) {
-            assertThat(bst.insert(e)).isFalse(); //does not insert
-            assertThat(bst.contains(e)).isTrue(); //but contains
+    void checkProperNodeFields() {
+        tree = new RecursiveBinarySearchTree<>();
+        Class<?> innerClass = getInnerClass();
+
+        boolean isElement = Arrays.stream(innerClass.getDeclaredFields())
+                .anyMatch(ELEMENT_FIELD);
+
+        boolean isLeft = Arrays.stream(innerClass.getDeclaredFields())
+                .anyMatch(LEFT_FIELD);
+
+        boolean isRight = Arrays.stream(innerClass.getDeclaredFields())
+                .anyMatch(RIGHT_FIELD);
+
+        assertThat(isElement).isTrue();
+        assertThat(isLeft).isTrue();
+        assertThat(isRight).isTrue();
+    }
+
+    @Test
+    @Order(4)
+    void shouldCreateTreeFromListOfElements() {
+        tree = RecursiveBinarySearchTree.of(someElements);
+        for (var e : someElements) {
+            assertThat(contains(getRootObject(), e)).isTrue();
         }
-        assertThat(bst.size()).isEqualTo(someElements.length);
+
+        assertThat(getInnerSize()).isEqualTo(someElements.length);
     }
 
+    @Test
+    @Order(5)
+    void shouldThrowNullPointerExceptionIfInsertNull() {
```

Comment:
@erneman1 according to the pattern we use, it should be called
`insertThrowsExceptionWhenArgumentIsNull`